### PR TITLE
Un-break LibreSSL

### PR DESCRIPTION
**Pull request by fboundp**
Commit d3e4424 addressing issue #64 caused code paths necessary for the
correct functioning of LibreSSL to be skipped. In particular:
 - Locking callbacks are no longer set, causing crashes when CL+SSL
   is used in an application with multiple threads.
 - SSL_library_init and SSL_load_error_strings are no longer called.
 - SSL_CTX_set_tmp_rsa_callback is no longer called (probably harmless,
   as this is only relevant if one is using export grade ciphersuites).

This was an easy to make oversight. The code paths above are
predicated on COMPAT-OPENSSL-VERSION being below version 1.1, since
they are no longer necessary for versions 1.1 or later. However,
LibreSSL APIs are compatible with version OpenSSL 1.0 APIs, yet the
version number is encoded as 2.0 (technically, #x20000000).

This change updates the conditional using LIBRESSLP, so that the
necessary code paths are restored. Additionally, it adds
OPENSSL-INIT-SSL and conditionally calls it instead of
SSL-INIT-LIBRARY for OpenSSL 1.1, where OPENSSL_init_ssl supersedes
SSL_library_init and SSL_load_error_strings.

### CODE REVIEW

**Comment by fboundp** [comment]
Giving Travis another go. Cannot reproduce failing build & test combinations locally on Ubuntu. ¯\_(ツ) _/¯

**Comment by fboundp** [comment]
I dunno what to think.
https://gist.github.com/fboundp/6650d42f262e7a69720a81375c2c1d3e

**Comment by avodonosov** [comment]
I restarted the master version Travis build, to ensure the failures are absent there: https://travis-ci.org/cl-plus-ssl/cl-plus-ssl/builds/594381622

However, the main thing I don't understand: you can see in the code usage of (libresslp). People have  suggested this implementation and confirmed it works: https://github.com/cl-plus-ssl/cl-plus-ssl/pull/79

Why it worked for others and doesn't work for you?

cc @iarebatman, @VaclavSynacek

**Comment by fboundp** [comment]
I suspect you will find that the failures are absent on master. I just have not yet been able to reproduce the failures seen with ECL and OpenSSL 1.1.*.

Regarding #79, I couldn’t say. Perhaps it was not tested on macOS, which is where I encountered the crash. CL+SSL prior to integrating #64 work just fine on recent major versions of macOS, but crash after #64 was merged (found via `git bisect`).

**Comment by avodonosov** [comment]
See my code review comment about the possible reason of the failures.

Please also check if adding '(:library libcrypto) to foreign function declarations as suggested in #88.
Maybe the reason for the problems you have is two foreign libraries loaded simultaneously: libressl and openssl. If not, we need to find the explanation of #79 working for others and not working for you.

**Comment by fboundp** [comment]
Woke up this morning with a hunch that played out. `OPENSSL_init_ssl` is not a one-for-one replacement for `SSL_library_init`. The former takes arguments 🤪. So I was getting lucky with stack layout with most of the matrix. Since docs say `OPENSSL_init_ssl` will be invoked automatically when needed, I simply removed that from the patch.

As to “we need to find the explanation of #79 working for others and not working for you”: I don’t believe it is working for others on macOS.

**Comment by fboundp** [comment]
(P.S. I cannot seem to find “code review comment about the possible reason of the failures”.)

**Comment by avodonosov** [comment]
https://github.com/cl-plus-ssl/cl-plus-ssl/pull/91#issuecomment-539283248

**Comment by avodonosov** [comment]
Probably until reviewer presses "submitvreview" others dont see comments

**Comment by avodonosov** [comment]
> I don’t believe it is working for others on macOS

We need to understand what makes macOS different.

Please also try :library libcypto

File: src/ffi.lisp [hunk 1]
```diff
@@ -282,6 +282,8 @@ Note: the _really_ old formats (<= 0.9.4) are not supported."
     :void)
 (define-ssl-function-ex (:vanished "1.1.0") ("SSL_library_init" ssl-library-init)
     :int)
+(define-ssl-function-ex (:since "1.1.0") ("OPENSSL_init_ssl" openssl-init-ssl)
+    :int)
```
[the remark this answers is not in the document]
Comment (fboundp) [reply src/ffi.lisp]: Haha! Yes, that’s exactly correct, as I also realized yesterday. That was removed from latest branch.

**Comment by fboundp** [comment]
I feel quite sure in this case I do not have multiple ssl/crypto libraries loaded … first thing I checked. (Verified using examination via vmmap, and also using DYLD_PRINT_LIBRARIES and DYLD_PRINT_APIS.) But I’ll give `:library crypto` a try … changed the `#+(and lispworks darwin)` to just `#+darwin` in that case, same results.

```console
env ROSWELL_HOME=/tmp/ros ros -e '(asdf:load-system :cl+ssl) (cl+ssl::ensure-initialized)'                                                                     ✗ 1
WARNING: System definition file #P"/private/tmp/ros/lisp/quicklisp/dists/quicklisp/software/flexi-streams-20190107-git/flexi-streams.asd" contains definition for system "flexi-streams-test". Please only define "flexi-streams" and secondary systems with a name starting with "flexi-streams/" (e.g. "flexi-streams/test") in that file.
CORRUPTION WARNING in SBCL pid 91629(tid 0x341adc0):
Memory fault at 0x18 (pc=0x7fff6660cefe, fp=0x4ff510, sp=0x4ff510) tid 0x341adc0
The integrity of this image is possibly compromised.
Continuing with fingers crossed.

debugger invoked on a SB-SYS:MEMORY-FAULT-ERROR in thread
#<THREAD "main thread" RUNNING {10005084C3}>:
  Unhandled memory fault at #x18.

Type HELP for debugger help, or (SB-EXT:EXIT) to exit from SBCL.

restarts (invokable by number or by possibly-abbreviated name):
  0: [CONTINUE] Ignore runtime option --eval "(ros:run '((:eval\"(ros:quicklisp)\")(:eval \"(asdf:load-system :cl+ssl) (cl+ssl::ensure-initialized)\")(:quit)))".
  1: [ABORT   ] Skip rest of --eval and --load options.
  2:            Skip to toplevel READ/EVAL/PRINT loop.
  3: [EXIT    ] Exit SBCL (calling #'EXIT, killing the process).

("bogus stack frame")
```

**Comment by fboundp** [comment]
OK, did some more homework for you. The crash occurs on macOS because:

1. Due to the change in #64, `SSL_library_init` is no longer called for LibreSSL, and thus algorithm tables are not populated.
2. The default loaded libssl, `/usr/lib/libssl.dylib`, on macOS is LibreSSL 2.2.7 (since macOS versions 10.12)
3. Prior to LibreSSL version 2.5.2 [^1], some digest methods for SSL (namely MD5 and SHA-1) are looked up during `SSL_CTX_new`. If the algorithm tables were not already loaded, this would cause `SSL_CTX_new` to fail.
4. CL+SSL does not check for failure of `SSL_CTX_new`, which leads to a null pointer dereference.

[^1] https://github.com/libressl-portable/openbsd/commit/b1dc16e9fa44da632faca9e01302db2b13513e71

**Comment by fboundp** [comment]
I pushed an update to the branch with updated description and tweak to make the diff more obvious.

**Comment by fboundp** [comment]
For the records.

| Path | earliest macOS version | LibreSSL version |
|-|-|-|
|/usr/lib/libssl.dylib| macOS Sierra 10.12 | LibreSSL 2.2.7|
|/usr/lib/libssl.35.dylib| macOS Sierra 10.12 | LibreSSL 2.2.7|
|/usr/lib/libssl.43.dylib|macOS High Sierra 10.13 | LibreSSL 2.5.5 |
|/usr/lib/libssl.44.dylib |macOS Mojave 10.14 | LibreSSL 2.6.5 |
|/usr/lib/libssl.46.dylib|macOS Catalina 10.15 | LibreSSL 2.8.3 |

Probably the library order for `:darwin` in `src/reload.lisp` ought to try specific versions, to avoid problems like GH-56. This branch does not address this.

**Comment by slyrus** [comment]
Any update on this? Will this land soon? Would be great to fix my cl+ssl/abcl issue.

**Comment by avodonosov** [comment]
@slyrus, but this is a different issue?

I hope to commit yours today..

**Comment by avodonosov** [comment]
@fboundp, do you mean /usr/lib/ normally contains several versions of libressl on mac?

(I would expect a single libssl.N.dylib and libssl.dylib as a symlink to it).

**Comment by avodonosov** [comment]
@fboundp, if you're open to contributing more, consider adding LibreSSL to CI, similar to what we have now for various OpenSSL versions.

**Comment by fboundp** [comment]
> @fboundp, do you mean /usr/lib/ normally contains several versions of libressl on mac?
> 
> (I would expect a single libssl.N.dylib and libssl.dylib as a symlink to it).

Correct, /usr/lib contains multiple versions. /usr/lib/libssl.dylib is actually a “stub” library that is itself linked to and re-exports /usr/lib/libssl.35.dylib. It appears to send telemetry back to Apple when it is loaded, with an event tagged `com.apple.libressl.unversioned_load`. There is also the following string, included in a crash log when running software built against the macOS 10.15 SDK.

```Invalid dylib load. Clients should not load the unversioned libssl dylib as it does not have a stable ABI.```

and another string that I can’t quite figure out how it gets triggered:

```WARNING: Executing a script that is loading libssl in an unsafe way. This will fail in a future version of macOS. Set the LIBRESSL_REDIRECT_STUB_ABORT=1 in the environment to force this into an error.```

It’s pretty weird. But clearly, apps are meant to use the versioned libssl libraries at runtime.

**Pull request closed without merging**